design-proposals/golden-image-delivery-and-update-pipeline.md

```diff
@@ -0,0 +1,207 @@
+
+These VolumeImportPollers can be configured on the SSP operators Custom Resource.
+
+
+![SSP Disk Import Pipelines](golden-image-delivery-images/ssp-disk-import-pipelines.png "SSP Disk Import Pipelines")
+
+
+## Common templates: VolumeReferences as DataVolume Source
+
+The common templates shipped with SSP need to have their DataVolumeTemplate sections transition to using a VolumeReference as the source field rather than a PVC claim name. This allows the VolumeImportPoller to seamlessly update the underlying PVC the corresponding VolumeReference points to, which then ensures that the common templates are always using the most up-to-date disk images.
+
+## HCO: Integration with SSP Operator
+
+The HCO is responsible for setting VolumeImportPoller sources on the SSP operator’s CR. These volume sources are the locations used by the VolumeImportPollers to populate PVCs for the common templates.
+
+
+## OCP UI: Integration with VolumeReferences
+
+The UI currently detects if a template has a boot source by checking if the DataVolumeTemplate’s source PVC is present. The UI needs to now be capable of checking if a DataVolumeTemplate’s source VolumeReference has phase: linked as well in order to determine if a boot source is set.
+
+## Scalability
+
+The only scaling concern is that the automated process has the potential to exhaust storage resources unless a PVC garbage collection mechanism is observed by the VolumeImportPoller.
+
+## Update/Rollback Compatibility
+
+* SSP operator must make image pipelines an opt-in feature to ensure existing boot images are not overwritten. 
+* UI must be capable of detecting boot sources for DataVolumeTemplates with sources for both VolumeReferences and PVCs
+
+## Implementation Phases
+
+**Phase1** - CDI APIs
+* VolumeReference API/controller
+* VolumeImportPoller API/controller
+
+**Phase2** - SSP + Common Templates
+* Common Templates transition to VolumeReferences for DataVolumeTemplate source
+* SSP Operator creates VolumeImportPollers for automatic PVC import
+
+**Phase3** - HCO and UI
+* HCO Team integrates with SSP Operator’s API for setting volume import sources
+* UI Team transitions to detecting boot source using a linked VolumeReference rather than looking directly for a PVC of a specific name.
+
+
+
+# API Examples
+## VM + DataVolumeTemplate + VolumeReference as source
+
+```yaml
+apiVersion: cdi.kubevirt.io/v1beta1
+kind: DataVolume
+metadata:
+   name: my-fedora-vm-image
+spec:
+   pvc:
+     accessModes:
+     - ReadWriteOnce
+     resources:
+       requests:
+         storage: 2Gi
+   source:
+     volumeReference:
+       name: fedora
+       namespace: some-namespace
+```
+
+## VolumeImportPoller Example
+
+This example polls for image tags for the quay.io/kubevirt/fedora image and imports the latest image sorting tags based on semver. As new tags are introduced, the VolumeImportPoller will import them as new PVCs and automatically manage updating the corresponding VolumeReference.
```

Comment:
I'd probably allow specifying a tag directly and always polling for the newes sha instead.
This way it's possible to specify the e.g. 4.8 tag and always have the newest 4.8 image (4.8.9...).

Reply:
> I'd probably allow specifying a tag directly and always polling for the newes sha instead.

Maybe that makes sense someday, but I don't think we're interested in using floating tags at the moment. In general, I'm not sure if we'd ever consider going into production with a floating tag either.

The primary concern with a SHA alone is that it's difficult to tie a SHA to an exact image version.  If someone said they're using centos8 with some SHA, i'd have no idea what version of centos 8 that is unless i downloaded and booted the image myself.  If someone said their using `centos8:3.2011` then I know exactly what's going on. 

The approach i've outlined here appears to align with the fedora automation [1] for publishing container disks. It posts the fedora:33 image once, then the fedora:34 image when it comes out, then the fedora:35 image in the future when that one comes out. There is no "fedora:latest" floating tag being used.

A similar thing would go for something like centos 8. I'd expect a centos 8 container image with tags representing the x.z versions. The most recent release would be `centos8:3.2011`. The importer would index all the tags for the centos8 image and then import the highest semver. 

1. https://github.com/kubevirt/common-templates/blob/6e64be41089a6de810627d972287b3bc42726100/dvtemplates/fedora_new_image.sh#L48

Reply:
I was thinking if we'd want to offer avoiding major version upgrades, but makes a lot of sense this way. Thanks